adapters/repos/db/inverted/new_prop_length_tracker.go

```diff
@@ -0,0 +1,246 @@
+	"os"
+	"sync"
+
+	"github.com/pkg/errors"
+)
+
+type PropLenData struct {
+	BucketedData map[string]map[int]int
+}
+
+type JsonPropertyLengthTracker struct {
+	path string
+	data PropLenData
+	sync.Mutex
+}
+
+func NewJsonPropertyLengthTracker(path string) (*JsonPropertyLengthTracker, error) {
+	t := &JsonPropertyLengthTracker{
+		data: PropLenData{make(map[string]map[int]int)},
+		path: path,
+	}
+
+	// read the file into memory
+	bytes, err := os.ReadFile(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			t.Flush()
+			return t, nil
+		}
+		return nil, err
+	}
+	t.path = path
+
+	var data PropLenData
+	if err := json.Unmarshal(bytes, &data); err != nil {
+		if bytes[0] != '{' {
+			// It's probably the old format file, load the old format and convert it to the new format
+			plt, err := NewPropertyLengthTracker(path)
+			if err != nil {
+				return nil, errors.Wrap(err, "convert old property length tracker")
+			}
+
+			propertyNames := plt.PropertyNames()
+			data = PropLenData{make(map[string]map[int]int, len(propertyNames))}
+			// Loop over every page and bucket in the old tracker and add it to the new tracker
+			for _, name := range propertyNames {
+				data.BucketedData[name] = make(map[int]int, 64)
+				for i := 0; i <= 64; i++ {
+					fromBucket := i
+					if i == 64 {
+						fromBucket = -1
+					}
+					count, err := plt.BucketCount(name, uint16(fromBucket))
+					if err != nil {
+						return nil, errors.Wrap(err, "convert old property length tracker")
+					}
+					data.BucketedData[name][fromBucket] = int(count)
+				}
+			}
+			t.data = data
+			t.FlushBackup()
+			plt.Close()
+			plt.Drop()
+			t.Flush()
+		}
+	}
+	t.data = data
+
+	return t, nil
+}
+
+func (t *JsonPropertyLengthTracker) FileName() string {
+	return t.path
+}
+
+func (t *JsonPropertyLengthTracker) TrackProperty(propName string, value float32) error {
+	t.Lock()
+	defer t.Unlock()
+
+	bucketId := t.bucketFromValue(value)
+	if _, ok := t.data.BucketedData[propName]; ok {
+		t.data.BucketedData[propName][int(bucketId)] = t.data.BucketedData[propName][int(bucketId)] + 1
+	} else {
+
+		t.data.BucketedData[propName] = make(map[int]int, 64)
+		t.data.BucketedData[propName][int(bucketId)] = 1
+	}
+
+	return nil
+}
+
+func (t *JsonPropertyLengthTracker) bucketFromValue(value float32) int {
+	if value <= 5.00 {
+		return int(value) - 1
+	}
+
+	bucket := int(math.Log(float64(value)/4.0)/math.Log(1.25) + 4)
+	if bucket > 63 {
+		return 64
+	}
+	return int(bucket)
+}
+
+func (t *JsonPropertyLengthTracker) valueFromBucket(bucket int) float32 {
+	if bucket <= 5 {
+		return float32(bucket + 1)
+	}
+
+	return float32(4 * math.Pow(1.25, float64(bucket)-3.5))
+}
+
+func (t *JsonPropertyLengthTracker) PropertyMean(propName string) (float32, error) {
+	t.Lock()
+	defer t.Unlock()
+
+	bucket, ok := t.data.BucketedData[propName]
+	if !ok {
+		return 0, nil // Needed for backwards compatibility
+	}
+
+	sum := float32(0)
+	totalCount := float32(0)
+
+	for i := -1; i <= 64; i++ {
```

Comment:
the 64 seems to be a magic number, could you create a constant?

Reply:
why do we start at -1?

Reply:
The answer to these requires a small write-up, or blog post, or something.  Do we have a place to put these?  I feel like we should be writing some extra commentary alongside our work

Reply:
added comment